opencog/atomspace/TypeIndex.cc

```diff
@@ -48,6 +48,31 @@ void TypeIndex::removeAtom(AtomPtr a)
 	remove(t, a->getHandle());
 }
 
+size_t TypeIndex::getNumAtomsOfType(Type type, bool count_subclasses) const
+{
+	iterator it(type, count_subclasses);
+	size_t atom_count = 0;
+	it.s = idx.begin();
+	it.send = idx.end();
+
+	// Loop over all the types looking for type and optional subclasses.
+	Type current_type = 0;
+	while (it.s != it.send)
+	{
+		// If this type is a match...
+		if ((type == current_type) || 
+		    (count_subclasses && (classserver().isA(type, it.type))))
+		{
+			// Add the size of the atom set for this type.
+			atom_count += idx.at(current_type).size();
+		}
+		current_type++;
+		++it.s;
+	}
+
```

Comment:
The existing ClassServer vectors implement a two-dimensional matrix of size number_of_atoms which only contain the IsA relationship as a `std::vector< std::vector<bool> >` which is there for the array-style lookups in IsA. What I'm envisioning is a `vector< vector<Type> >` so the types can be directly iterated over without any boolean checks and skipping of all non-subclasses. Would be faster for iterating and slower for lookups of IsA so we really need both for the two different ways these types are accessed: both sequential and random.